### PR TITLE
Bert sequence parallel support

### PR DESCRIPTION
**Pull request by shanmugamr1992**
# What does this PR do ?

Add a one line overview of what this PR aims to accomplish.

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/nlp/models/language_modeling/megatron/bert_model.py [hunk 1]
```diff
@@ -71,14 +72,23 @@ class BertLMHead(MegatronModule):
...
-        self, mpu_vocab_size, hidden_size, init_method, layernorm_epsilon, parallel_output, use_openai_gelu, onnx_safe
+        self,
+        mpu_vocab_size,
+        hidden_size,
+        init_method,
+        layernorm_epsilon,
+        parallel_output,
+        use_openai_gelu,
+        onnx_safe,
+        sequence_parallel,
```
Comment (markelsanz14) [inline nemo/collections/nlp/models/language_modeling/megatron/bert_model.py:83]: Can you set default values for some of these parameters?

File: nemo/collections/nlp/models/language_modeling/megatron/bert_model.py [hunk 2]
```diff
@@ -155,15 +166,18 @@ def __init__(
         hidden_dropout=0.1,
         precision=16,
         fp32_residual_connection=False,
+        activations_checkpoint_granularity=None,
         activations_checkpoint_method=None,
         activations_checkpoint_num_layers=1,
+        activations_checkpoint_layers_per_pipeline=None,
```
Comment (markelsanz14) [inline nemo/collections/nlp/models/language_modeling/megatron/bert_model.py:172]: Shouldn't num_micro_batches_with_partial_activation_checkpoints also be set somewhere here? Not sure

File: nemo/collections/nlp/models/language_modeling/megatron/bert_model.py [hunk 3]
```diff
@@ -197,14 +212,17 @@ def __init__(
             use_cpu_initialization=use_cpu_initialization,
             precision=precision,
             fp32_residual_connection=fp32_residual_connection,
-            activations_checkpoint_method=activations_checkpoint_method,
-            activations_checkpoint_num_layers=activations_checkpoint_num_layers,
+            activations_checkpoint_granularity=None,
+            activations_checkpoint_method=None,
+            activations_checkpoint_num_layers=1,
+            activations_checkpoint_layers_per_pipeline=None,
```
Comment (markelsanz14) [inline nemo/collections/nlp/models/language_modeling/megatron/bert_model.py:218]: Shouldn't num_micro_batches_with_partial_activation_checkpoints also be set somewhere here? Not sure

File: nemo/collections/nlp/models/language_modeling/megatron/bert_model.py (hunk 2 above)
Comment (shanmugamr1992) [reply nemo/collections/nlp/models/language_modeling/megatron/bert_model.py:172]: Following what ever was on GPT3.

File: nemo/collections/nlp/models/language_modeling/megatron/bert_model.py (hunk 1 above)
Comment (shanmugamr1992) [reply nemo/collections/nlp/models/language_modeling/megatron/bert_model.py:83]: Yes, I am setting on sequence parallel alone, and leaving others just as was before.

File: nemo/collections/nlp/models/language_modeling/megatron/bert_model.py [hunk 4]
```diff
@@ -220,18 +238,29 @@ def __init__(
                 parallel_output,
                 openai_gelu,
                 onnx_safe,
+                sequence_parallel,
             )
             self._lm_head_key = 'lm_head'
             self.binary_head = None
             if self.add_binary_head:
+                """
+                Should we do something here when sequence parallel is enabled ?
```
Comment (yidong72) [inline nemo/collections/nlp/models/language_modeling/megatron/bert_model.py:247]: The pooler layer is doing the gather? so there is no more SP ? please verify and remove the comments

**Review by yidong72 — COMMENTED** [review]
Loos good. Could you also add the CI test?

File: nemo/collections/nlp/models/language_modeling/megatron/bert_model.py (hunk 4 above)
Comment (shanmugamr1992) [reply nemo/collections/nlp/models/language_modeling/megatron/bert_model.py:247]: Yes thats true. Checked. Will remove

File: nemo/collections/nlp/modules/common/megatron/language_model.py [hunk 5]
```diff
@@ -178,7 +178,7 @@ def forward(self, hidden_states, sequence_index=0):
         # gather data along sequence dimensions
         # same pooler is run on all tensor parallel nodes
         if self.sequence_parallel:
-            hidden_states = tensor_parallel.gather_from_sequence_parallel_region()
+            hidden_states = tensor_parallel.mappings.gather_from_sequence_parallel_region(hidden_states)
```
Comment (shanmugamr1992) [inline nemo/collections/nlp/modules/common/megatron/language_model.py:181]: @yidong72  Make sure this is okay

Comment (yidong72) [reply nemo/collections/nlp/modules/common/megatron/language_model.py:181]: looks good. please verify the shape before and after gather is the expected

**Review by yidong72 — APPROVED** [review]
LGTM.

**Pull request merged**